### ISSUE TITLE
Bug: `no-loop-func` errors on functions in the condition and not just the body

### ISSUE THREAD

**Issue by LJNeon** [eslint/eslint#19120]
### Environment

Node version: v22.11.0
npm version: v10.9.0
Local ESLint version: v9.14.0 (Currently used)
Global ESLint version: Not found
Operating System: linux 6.8.0-48-generic

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
export default [{rules: {"no-loop-func": 2}}];
```
</details>

```js
const array = [5, 7, 8, 3];
let idx = 0;

while(array.some(elm => elm > idx))
  idx += 1;
```

### What did you expect to happen?

No error reported as there are no unsafe references.

### What actually happened?

An error was reported, citing the arrow function inside the condition of the while statement.

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiY29uc3QgYXJyYXkgPSBbNSwgNywgOCwgM107XG5sZXQgaWR4ID0gMDtcblxud2hpbGUoYXJyYXkuc29tZShlbG0gPT4gZWxtID4gaWR4KSlcbiAgaWR4ICs9IDE7XG4iLCJvcHRpb25zIjp7InJ1bGVzIjp7Im5vLWxvb3AtZnVuYyI6WyJlcnJvciJdfSwibGFuZ3VhZ2VPcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

Functions inside of the condition are evaluated once per loop and are thus safe to reference variables that are modified. This is a false positive.

**Comment by Tanujkanti4441** [comment]
Hi, @LJNeon, thanks for the issue.

Seems like a bug to me, as its docs also says
> Disallow function declarations that contain unsafe references inside loop statements

so rule should not report the functions present in loop's condition.

@eslint/eslint-team, should we consider this?

**Comment by fasttime** [comment]
I believe this is not a bug: it's intended that functions trigger the rule also in a loop's condition. The example provided in the issue is equivalent to the following snippet, where the expression `array.some(elm => elm > idx)` is moved from the loop's condition into the body:

```js
const array = [5, 7, 8, 3];
let idx = 0;

while (true) {
  if (!array.some(elm => elm > idx)) break;
  idx += 1;
}
```
[**Playground Link**](https://eslint.org/play/#eyJ0ZXh0IjoiY29uc3QgYXJyYXkgPSBbNSwgNywgOCwgM107XG5sZXQgaWR4ID0gMDtcblxud2hpbGUgKHRydWUpIHtcbiAgaWYgKCFhcnJheS5zb21lKGVsbSA9PiBlbG0gPiBpZHgpKSBicmVhaztcbiAgaWR4ICs9IDE7XG59Iiwib3B0aW9ucyI6eyJydWxlcyI6eyJuby1sb29wLWZ1bmMiOlsiZXJyb3IiXX0sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19)

I don't see any unit tests with a function in a loop's condition (this could be improved). There is however a valid test case where a `for`-initializer contains a function, and in that case the rule correctly reports no problems:

https://github.com/eslint/eslint/blob/89a4a0a260b8eb11487fe3d5d4d80f4630933eb3/tests/lib/rules/no-loop-func.js#L25

**Comment by fasttime** [comment]
> Functions inside of the condition are evaluated once per loop and are thus safe to reference variables that are modified.

This isn't correct. In the code example provided in the issue, the function is called 16 times, see this modified snippet:

```js
const array = [5, 7, 8, 3];
let idx = 0;
let fnCount = 0;

while(array.some(elm => (fnCount++, elm > idx)))
  idx += 1;

console.log(fnCount); // prints 16
```

Whether and how many times the function is called depends on the surrounding expression, here `array.some(...)`.

**Comment by mdjermanovic** [comment]
I agree with @fasttime. Loop condition is part of the loop. In the example provided in the issue, a new instance of the function is created in each iteration, which is exactly what this rule targets. While this particular example is a false positive, I think it isn't a bug but a limitation of this rule as it can't know whether the function instance is just immediately invoked and then discarded, or possibly stored for later use. Overall, I think this is a case for using `eslint-disable` comments.

**Comment by Tanujkanti4441** [comment]
>I think it isn't a bug but a limitation of this rule as it can't know whether the function instance is just immediately invoked and then discarded, or possibly stored for later use. Overall, I think this is a case for using eslint-disable comments.

So, how about adding this info in the limitation section of the rule docs.

**Comment by fasttime** [comment]
Sorry for losing track of this. 

> > I think it isn't a bug but a limitation of this rule as it can't know whether the function instance is just immediately invoked and then discarded, or possibly stored for later use. Overall, I think this is a case for using eslint-disable comments.
> 
> So, how about adding this info in the limitation section of the rule docs.

I'd be also fine with this change. Marking as accepted since there is agreement on expanding the rule docs.
